compiler/src/dmd/file_manager.d

```diff
@@ -26,14 +26,86 @@ bool isPackageFileName(scope FileName fileName) nothrow
     return FileName.equals(fileName.name, package_d) || FileName.equals(fileName.name, package_di);
 }
 
+// A path stack that allows one to go up and down the path using directory
+// separators. `cur` is the current path, `up` goes up one path, `down` goes
+// down one path. if `up` or `down` return false, there are no further paths.
+private struct PathStack
+{
+    private const(char)[] path;
+    private size_t pos;
+
+    @safe @nogc nothrow pure:
+
+    this(const(char)[] p)
+    {
+        path = p;
+        pos = p.length;
+    }
+
+    const(char)[] cur()
+    {
+        return path[0 .. pos];
+    }
+
+    bool up()
+    {
+        if (pos == 0)
+            return false;
+        while (--pos != 0)
+            if (isDirSeparator(path[pos]))
+                return true;
+        return false;
+    }
+
+    bool down()
+    {
+        if (pos == path.length)
+            return false;
+        while (++pos != path.length)
+            if (isDirSeparator(path[pos]))
+                return true;
+        return false;
+    }
+}
+
 final class FileManager
 {
     private StringTable!(const(ubyte)[]) files;
+    private StringTable!(bool) packageStatus;
+
+    // check if the package path of the given path exists. The input path is
+    // expected to contain the full path to the module, so the parent
+    // directories of that path are checked.
+    private bool packageExists(const(char)[] p) nothrow
+    {
+        // step 1, look for the closest parent path that is cached
+        bool exists = true;
+        auto st = PathStack(p);
+        while (st.up) {
+            if (auto cached = packageStatus.lookup(st.cur)) {
+                exists = cached.value;
+                break;
+            }
+        }
+        // found a parent that is cached (or reached the end of the stack).
+        // step 2, traverse back up the stack storing either false if the
```

Comment:
This should be

"traverse back down"

right?

Reply:
Yeah could be. Not really great terminology.